### ISSUE TITLE
useEvent hook not working for custom UI modules

### ISSUE THREAD

**Issue by mrbrentkelly** [software-mansion/react-native-reanimated#1229]
## Description

The [useEvent](https://github.com/software-mansion/react-native-reanimated/blob/20ae770f73f65023278e7081ebe29d342a08c6e1/src/reanimated2/Hooks.js#L33) hook from Reanimated v2 doesn't work for events on a custom Android and iOS UI modules that use `RCTEventEmitter`. 

If events are emitted using the `EventDispatcher` from `UIManagerModule` then the `useEvent` hook works just fine.

**Works**

```java
EventDispatcher eventDispatcher = context
   .getNativeModule(UIManagerModule.class)
   .getEventDispatcher();

SomeCustomEvent event = SomeCustomEvent.obtain(/*params*/);

eventDispatcher.dispatchEvent(event);
```

**Doesn't work**

```java
context.getJSModule(RCTEventEmitter.class).receiveEvent(getId(), "SomeEvent", data);
```

The `RCTEventEmitter` approach is probably what most custom native modules use (for example [react-native-video](https://github.com/react-native-community/react-native-video/blob/master/android/src/main/java/com/brentvatne/react/ReactVideoView.java#L166)) since that's what the [docs](https://reactnative.dev/docs/native-components-android#events) say to use. So I think this might become a common gotcha if not fixed or at least documented.

> Note, this issue exists on iOS as well.

## Steps To Reproduce

1. Create new RN project and install reanimated2
2. Create a Native UI module on Android that emits events using RCTEventEmitter
3. Consume events on the JS layer using the `useEvent` hook.

### Expected behavior

Events triggered using RCTEventEmitter should be picked up by the `useEvent` hook

### Actual behavior

The worklet inside the `useEvent` hook is not triggered.

## Snack or minimal code example

I've created a new RN project that repro's the issue:

https://github.com/mrbrentkelly/reanimated2-event-bug

<img src="https://user-images.githubusercontent.com/5814579/93265805-0b52ea00-f7a1-11ea-829c-417767312d95.gif" width="200" />

## Package versions

- React: 16.13.1
- React Native: 0.63.2
- React Native Reanimated: 2.0.0

**Comment by jakub-gonet** [comment]
Sorry for the late response.

Are those events direct (sent to the `View`s)? AFAIK we can only handle that type of events.

**Comment by mrbrentkelly** [comment]
Yes these are direct view events (although I would expect RE2 to work with bubbling view events also). 

See my example [here](https://github.com/mrbrentkelly/reanimated2-event-bug/blob/master/android/app/src/main/java/com/reanimatedeventbug/MyCustomViewManager.java#L51) 👍

**Comment by mrbrentkelly** [comment]
Just verified that this issue is occurring on iOS also... i.e. The only way for the reanimated2 `useEvent` hook to pick up events from a custom native module is to use the `RCTEventDispatcher`, which is not a common way to implement events on iOS UI Modules.

Works:

```swift
// Native View in a custom ViewManager

let eventDispatcher = bridge.eventDispatcher()
eventDispatcher.send(OnSomeDirectEvent(viewTag: viewTag, data))
eventDispatcher.send(OnSomeBubblingEvent(viewTag: viewTag, data))
```

Does not work:

```swift
// Native View in a custom ViewManager

@objc public var onSomeDirectEvent: RCTDirectEventBlock?
@objc public var onSomeBubblingEvent: RCTDirectEventBlock?

onSomeDirectEvent?(data)
onSomeBubblingEvent?(data)
```

Are there plans for renimated2 to support UI component events that were implemented using the standard [UI component event mechanism](https://reactnative.dev/docs/native-components-ios#events)? Implementing events using the event dispatcher is much more involved as you have to create a custom `RCTEvent` class for each event and manage coalescing.

**Comment by terrysahaidak** [comment]
Hi, I'm pretty sure reanimated would want to support custom ui events but it's just not a priority right now and team is working on something else.

If you could work on this PR I'm sure @Szymon20000 or @karol-bisztyga would provide some help and feedback.

**Comment by mrbrentkelly** [comment]
Sure thing 👍  Any directions to code on where I should start with Reanimated2 event logic would be appreciated and I can take a look. 

Hopefully this will be a non trivial change since its something that was supported in Reanimated1 with `Animated.Event`. I pretty much can't migrate any of my Reanimated1 code that's using `Animated.Event()` to `useEvent()` 💔.

**Comment by szydlovsky** [comment]
Closing the issue since the PR got already merged way back. If anything related to it still pops up please create new issues!

**Comment by mrbrentkelly** [comment]
> Closing the issue since the PR got already merged way back

@szydlovsky what was the PR that resolved this issue? Could you link to it?

**Comment by szydlovsky** [comment]
@mrbrentkelly I thought that https://github.com/software-mansion/react-native-reanimated/pull/2207 solved it. We are doing some older issues cleanup, sorry for making a hasty decision 😄

**Comment by mrbrentkelly** [comment]
No problem! Unfortunately I don't believe https://github.com/software-mansion/react-native-reanimated/pull/2207 tackled this particular issue, so I'd recommend we keep it open 👍
